File: leetcode/algorithms/703_kth_largest_element_in_a_stream/main.cpp
#include <queue>
#include <vector>
using namespace std;

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

class KthLargest {
private:
    priority_queue<int, vector<int>, greater<int>> _priorityQueue;
    int _k;
    
public:
    KthLargest(int k, vector<int>& nums) {
        _k = k;
        for (int i = 0; i < nums.size(); i++) {
            _priorityQueue.push(nums[i]);
            if (_priorityQueue.size() > k) {
                _priorityQueue.pop();
            }
        }
    }
    
    int add(int val) {
        _priorityQueue.push(val);
        if (_priorityQueue.size() > _k) {
            _priorityQueue.pop();
        }
        return _priorityQueue.top();
    }
};


// Solution
class Solution_KthLargest {
    priority_queue<int, vector<int>, greater<int>> pq;
    int size;

    Solution_KthLargest(int k, vector<int> nums) {
        size = k;
        for (int i = 0; i < nums.size(); i++) {
            pq.push(nums[i]);
            if (pq.size() > k) {
                pq.pop();
            }
        }
    }

    int add(int val) {
        pq.push(val);
        if (pq.size() > size) {
            pq.pop();
        }
        return pq.top();
    }
};

File: leetcode/algorithms/703_kth_largest_element_in_a_stream/main.py
import heapq
from queue import PriorityQueue
from typing import List

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)


class KthLargest:
    priorityQueue: PriorityQueue
    k: int

    def __init__(self, k: int, nums: List[int]):
        self.priorityQueue = PriorityQueue(k)
        self.k = k
        for num in nums:
            self.add(num)

    def add(self, val: int) -> int:
        if self.priorityQueue.qsize() < self.k:
            self.priorityQueue.put(val)
        else:
            top: int = self.priorityQueue.get(0)
            if top < val:
                self.priorityQueue.put(val)
            else:
                self.priorityQueue.put(top)

        top: int = self.priorityQueue.get(0)
        self.priorityQueue.put(top)

        return top


# Solution
class Solution_KthLargest(object):
    def __init__(self, k, nums):
        self.pool = nums
        self.k = k
        heapq.heapify(self.pool)
        while len(self.pool) > k:
            heapq.heappop(self.pool)

    def add(self, val):
        if len(self.pool) < self.k:
            heapq.heappush(self.pool, val)
        elif val > self.pool[0]:
            heapq.heapreplace(self.pool, val)
        return self.pool[0]
